File: Rust/sorting/insertion_sort.rs
/// Insertion Sort implementation
///
/// # Description
/// Insertion Sort works by building a sorted portion of the array one element at a time.
/// It is efficient for small or nearly sorted datasets.
///
/// # Time Complexity
/// - Best: O(n)
/// - Average/Worst: O(n²)
///
/// # Example
/// ```
/// let mut nums = vec![5, 2, 4, 6, 1, 3];
/// insertion_sort(&mut nums);
/// assert_eq!(nums, vec![1, 2, 3, 4, 5, 6]);
/// ```
pub fn insertion_sort(arr: &mut Vec<i32>) {
    for i in 1..arr.len() {
        let key = arr[i];
        let mut j = i;
        while j > 0 && arr[j - 1] > key {
            arr[j] = arr[j - 1];
            j -= 1;
        }
        arr[j] = key;
    }
}

File: TypeScript/sorting/simple/insertionSort.ts
/**
 * Insertion Sort Algorithm
  * 
 * Time Complexity: O(n²) worst case, O(n) best case
 * Space Complexity: O(1)
  * 
 * Builds the sorted array one item at a time by inserting each element
  * into its correct position.
 */
class InsertionSort {
  /**
   * Sorts an array using the Insertion Sort algorithm (ascending order).
   * 
   * @param arr Array of numbers to be sorted
   * @return Sorted array
   */
  static sort(arr: number[]): number[] {
    const n = arr.length;

    for (let i = 1; i < n; i++) {
      const key = arr[i];
      let j = i - 1;

      // Move elements greater than key one position ahead
      while (j >= 0 && arr[j] > key) {
        arr[j + 1] = arr[j];
        j--;
      }

      arr[j + 1] = key;
    }

    return arr;
  }

  /**
   * Sorts an array in descending order.
   */
  static sortDescending(arr: number[]): number[] {
    const n = arr.length;

    for (let i = 1; i < n; i++) {
      const key = arr[i];
      let j = i - 1;

      while (j >= 0 && arr[j] < key) {
        arr[j + 1] = arr[j];
        j--;
      }

      arr[j + 1] = key;
    }

    return arr;
  }
}
